### PR TITLE
fmt tool improvements

### LINKED ISSUE #898: sixtyfps-fmt formats to invalid syntax
Kind of a followup from https://github.com/sixtyfpsui/sixtyfps/discussions/895#discussioncomment-2070964

I noticed that the formatter is pretty mvp and gives sometimes incorrect behaviour, see example gif below.

CleanShot 2022-01-29 at 15 32 43

Would you mind me looking into this? The formatter tool seems to be a relatively small project, so it should not be much of a trouble.

-- side question 1
Is it a conscious decision to put no space after the component name and before { ? (see below) If not, can we change it to put space there?
CleanShot 2022-01-29 at 15 23 33@2x

-- side question 2
I guess in general I'd like to know if you are willing to review/accept some PRs related to the formatter..
Maybe it could get a README, tests, vscode extension, config, ...

Thanks! :)

### PR DESCRIPTION
**Pull request by Zajozor**
Closes #898 
Sorry it took a while - but i needed to learn more than anticipated :)

Changes in this PR:
- **unrelated to fmt** - I own an M1 Mac and the builds of most of the creates were failing for me unless I updated neon. I am not sure about the impact on this on the rest of the project though. Let me know if it breaks something and I'll remove the commit.

- Adds tests to fmt, with some basic examples
- Fixes the original issue from #898 as well as some things about BinaryExpression/Expression formatting
- Some code cleanup
- Adds a README with a tutorial on how to get this running in your vscode

Overall fmt is still far form perfect but I see this as a step in a good direction.

btw sorry if some of my changes still refer to the old name, you literally renamed 3 hours ago 😅

### CODE REVIEW

File: tools/fmt/main.rs [hunk 1]
```diff
@@ -134,63 +136,47 @@ fn process_markdown_file(source: String, mut file: impl Write) -> std::io::Resul
...
-type State = ();
+fn process_file(
+    source: String,
+    path: std::path::PathBuf,
+    mut file: impl Write,
+) -> std::io::Result<()> {
+    match path.extension() {
+        Some(ext) if ext == "rs" => return process_rust_file(source, file),
+        Some(ext) if ext == "md" => return process_markdown_file(source, file),
+        Some(ext) if ext == "60" => return process_60_file(source, path, file),
```
Comment (Zajozor) [inline tools/fmt/main.rs:163]: so actually are the files now also called .slint ? i'll look into renaming these in the next days

File: tools/fmt/writer.rs [hunk 2]
```diff
@@ -0,0 +1,37 @@
+// Copyright © SixtyFPS GmbH <info@sixtyfps.io>
+// SPDX-License-Identifier: (GPL-3.0-only OR LicenseRef-SixtyFPS-commercial)
```
Comment (ogoffart) [inline tools/fmt/writer.rs:2]: The CI system is checking the license header on every file, and that is the old one.

```suggestion
// Copyright © SixtyFPS GmbH <info@slint-ui.com>
// SPDX-License-Identifier: GPL-3.0-only OR LicenseRef-Slint-commercial
```

**Review by ogoffart — COMMENTED** [review]
Thank you very much. This look s like a great improvement

Yeah, sorry that you were disrupted by the naming change. There is still quite a lot of references to .60 and SixtyFPS, but i can as well fix them after the merge of this PR.

However let's just try to make the CI green before merging it.

**Comment by Zajozor** [comment]
You were right with the node16 thing, i was using node17 before, that was the problem, didn't realize :)

I removed the commit which upgraded neon and also renamed all the references of sixtyfps to slint.

Let me know if there is something else that needs to be done before merging! (from my pov it's ready)

**Comment by ogoffart** [comment]
Looks good! Thank you for the contribution

**Pull request merged**